[DLPack] Add support for missing keyword-arguments.

ysiraichi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #150691
* __->__ #150218
* #150217
* #150216

This PR introduces the rest of the keyword-arguments added in DLPack
version 2023.12: `dl_device` and `copy`.

In summary, we handle these arguments in the C++ implementation of
`to_dlpack(...)` at _torch/csrc/Module.cpp_, by calling the
`maybeCopyTensor` function at _aten/src/ATen/DLConvertor.cpp_. It also
introduces the following changes:

- Add a new Python API `torchDeviceToDLDevice()`, which is simply a
  refactoring of the `getDLDevice()` function at
  _aten/src/ATen/DLConvertor.cpp_.
- Add both keyword-arguments to the `from_dlpack()` function at
  _torch/utils/dlpack.py_ and to the `Tensor.__dlpack__()` dunder
  method.

ysiraichi committed:
Update

[ghstack-poisoned]

ysiraichi committed:
Update

[ghstack-poisoned]

ysiraichi committed:
Rebased.

[ghstack-poisoned]

ysiraichi committed:
Update

[ghstack-poisoned]

ysiraichi committed:
Update

[ghstack-poisoned]

ysiraichi committed:
Update

[ghstack-poisoned]

ysiraichi committed:
Update

[ghstack-poisoned]

ysiraichi committed:
Update

[ghstack-poisoned]

ysiraichi committed:
Update

[ghstack-poisoned]

ysiraichi committed:
Update

[ghstack-poisoned]

msaroufim commented on aten/src/ATen/DLConvertor.cpp:
```diff
@@ -388,4 +389,35 @@ Tensor fromDLPackVersioned(DLManagedTensorVersioned* src, std::function<void(voi
+          data.device(),
+          " to ",
+          device,
+          " without copying. Set copy=True is needed.");
```
just double checking this check would make it clear that users would need to set the copy flag in

`torch.from_dlpack(..., copy=copy)`

albanD replied:
copy=None would also work. It's more that you cannot set copy=False in this case

albanD commented on torch/utils/dlpack.py:
```diff
@@ -54,7 +54,7 @@ class DLDeviceType(enum.IntEnum):
 # TODO: add a typing.Protocol to be able to tell Mypy that only objects with
 # __dlpack__ and __dlpack_device__ methods are accepted.
-def from_dlpack(ext_tensor: Any) -> 'torch.Tensor':
+def from_dlpack(ext_tensor: Any, device: Any = None, copy: Optional[bool] = None) -> 'torch.Tensor':
```
kwarg-only!

albanD replied:
Docstring needs updating.
Also is the "device" type really any? We have  Device in our torch.types utils if needed.

albanD commented on torch/csrc/Module.cpp:
```diff
@@ -636,6 +668,22 @@ static PyObject* THPModule_fromDLPack(PyObject* _unused, PyObject* data) {
+  auto device = reinterpret_cast<THPDevice*>(data)->device;
+  auto dl_device = at::torchDeviceToDLDevice(device);
+  auto tuple = PyTuple_New(2);
+  PyTuple_SetItem(tuple, 0, THPUtils_packInt64(dl_device.device_type));
```
nit here and below for perf
```suggestion
  PyTuple_SET_ITEM(tuple, 0, THPUtils_packInt64(dl_device.device_type));
```

albanD commented on torch/csrc/Module.cpp:
```diff
@@ -636,6 +668,22 @@ static PyObject* THPModule_fromDLPack(PyObject* _unused, PyObject* data) {
+      "torchDeviceToDLDevice: expected torch.device argument.");
+  auto device = reinterpret_cast<THPDevice*>(data)->device;
+  auto dl_device = at::torchDeviceToDLDevice(device);
+  auto tuple = PyTuple_New(2);
```
Add error checking in case this failed

ysiraichi replied:
Not sure I get what you mean. If `at::torchDeviceToDLDevice` errors, the error will bubble up to Python, won't it?

albanD replied:
CPython is C API, not C++. There is no error throwing here, you need to check the tuple is not null and if so, make this a c++ error by `throw python_error();` 

```cpp
if (!tuple) {
  throw python_error();
}
```

albanD commented on aten/src/ATen/DLConvertor.cpp:
```diff
@@ -96,10 +96,14 @@ DLDataType getDLDataType(const Tensor& t) {
-  ctx.device_id = static_cast<int32_t>(static_cast<unsigned char>(device_id));
-  switch (tensor.device().type()) {
+
+  ctx.device_id = (device.is_cuda() || device.is_privateuseone())
```
Why this new logic? We would we ignore the device index being passed in?

ysiraichi replied:
I just got it from the place it was being called before. Needed it, since I'm also calling this function outside.

ysiraichi replied:
Previously, this is how we were calling this function (the only call site):

https://github.com/pytorch/pytorch/blob/e1f28fe17bc55389fc5a31bad588d22e5ecca722/aten/src/ATen/DLConvertor.cpp#L317-L321

So, I basically, moved that code into this function. In other words, the logic did not change. The device index will always be either `src.get_device()`, i.e. `src.device().index()` (if the condition `src.is_cuda() || src.is_privateuseone()` is true), or 0 (otherwise).

albanD replied:
Ok, I guess it is another case where we should have an issue and tag other accelerator backends.

albanD commented on aten/src/ATen/DLConvertor.cpp:
```diff
@@ -130,38 +133,40 @@ static DLDevice getDLDevice(const Tensor& tensor, c10::DeviceIndex device_id) {
+      return at::Device(DeviceType::HIP, index);
 #endif
     case DLDeviceType::kDLOneAPI:
+      TORCH_CHECK(data != nullptr, "Can't get ATen device for XPU without XPU data.");
```
nit: Why do we use data here and not the passed in type and index?

ysiraichi replied:
I'm not really sure myself. I just added this check because I needed to use this function without `data` when copying the tensor to some other device (see `maybeCopyTensor`).

guangyey replied:
Currently, XPU always gets its device index from data since device index in PyTorch might not match the order of the xpu driver enumerated.
We will refine this issue in the following PR when the appropriate xpu toolchain is available.
BTW, here `TORCH_CHECK` is not needed since `getDeviceFromPtr` will check data's validity inside itself.

albanD commented on aten/src/ATen/DLConvertor.cpp:
```diff
@@ -388,4 +389,35 @@ Tensor fromDLPackVersioned(DLManagedTensorVersioned* src, std::function<void(voi
+  }
+
+  if (force_copy) {
+    return data.detach().clone();
```
There should be no autograd here already right? So detach() is not needed

albanD commented on torch/_tensor.py:
```diff
@@ -1673,7 +1673,13 @@ def __torch_function__(cls, func, types, args=(), kwargs=None):
-    def __dlpack__(self, stream=None, max_version=None):
+    def __dlpack__(
+        self,
+        stream: Optional[Any] = None,
```
kwargs only

albanD commented on torch/_tensor.py:
```diff
@@ -1673,7 +1673,13 @@ def __torch_function__(cls, func, types, args=(), kwargs=None):
+        stream: Optional[Any] = None,
+        max_version: Optional[tuple[int, int]] = None,
+        dl_device: Optional[tuple[enum.IntEnum, int]] = None,
+        copy: Optional[bool] = None,
```
docstring update

albanD commented on torch/csrc/Module.cpp:
```diff
@@ -607,25 +607,57 @@ void DLPack_Capsule_Destructor(PyObject* data) {
+        tensor, at::DLPackTraits<T>::capsule, DLPack_Capsule_Destructor<T>);
+  }
+
+  return nullptr;
```
I'm don't recall, will this have the right python err set?

ysiraichi replied:
I think so. At least, it will throw an error inside the parser. I will replace it with `Py_RETURN_NONE`, just for consistency.

albanD commented on torch/utils/dlpack.py:
```diff
@@ -107,19 +107,34 @@ def from_dlpack(ext_tensor: Any) -> 'torch.Tensor':
+        kwargs["dl_device"] = device
+
+        ext_device = ext_tensor.__dlpack_device__()
+        # ext_device is either CUDA or ROCm, we need to pass the current
```
Note that a lot of this string handling should get generalized to any device which is an accelerator in PT.
Ok to do later, but as we're going to see issues with hip or xpu, we should just refactor it together

ysiraichi replied:
Could you help me understand what string handling you are referring to? Doesn't `torch.device(device)` work with any accelerator in PT?

albanD replied:
Only the fact that you do custom processing only for CUDA and ROCm but we have quite a few other accelerators that would require the same treatment in the furure.

ysiraichi replied:
Since I'm not familiar with these other accelerators, I think I will leave it for a future PR. In this one, I'm simply adding support for the extra keywords (in this case, passing them through to `__dlpack__()` method). What do you think?

albanD replied:
Yes let's just open an issue for it and add the xpu and privateuse1 labels on it.

albanD commented on torch/utils/dlpack.py:
```diff
@@ -129,11 +144,18 @@ def from_dlpack(ext_tensor: Any) -> 'torch.Tensor':
+            # corresponding to DLPack 2022.12 specification.
             kwargs.pop("max_version")
+            kwargs.pop("dl_device")
+            kwargs.pop("copy")
```
Is it actually ok to ignore these arguments?!

ysiraichi committed:
Update

[ghstack-poisoned]

ysiraichi committed:
Rebased.

[ghstack-poisoned]

ysiraichi committed:
Update

[ghstack-poisoned]

ysiraichi committed:
Update on "[DLPack] Add support for missing keyword-arguments."


This PR introduces the rest of the keyword-arguments added in DLPack
version 2023.12: `dl_device` and `copy`.

In summary, we handle these arguments in the C++ implementation of
`to_dlpack(...)` at _torch/csrc/Module.cpp_, by calling the
`maybeCopyTensor` function at _aten/src/ATen/DLConvertor.cpp_. It also
introduces the following changes:

- Add a new Python API `torchDeviceToDLDevice()`, which is simply a
  refactoring of the `getDLDevice()` function at
  _aten/src/ATen/DLConvertor.cpp_.
- Add both keyword-arguments to the `from_dlpack()` function at
  _torch/utils/dlpack.py_ and to the `Tensor.__dlpack__()` dunder
  method.

[ghstack-poisoned]

ysiraichi committed:
Update on "[DLPack] Add support for missing keyword-arguments."


This PR introduces the rest of the keyword-arguments added in DLPack
version 2023.12: `dl_device` and `copy`.

In summary, we handle these arguments in the C++ implementation of
`to_dlpack(...)` at _torch/csrc/Module.cpp_, by calling the
`maybeCopyTensor` function at _aten/src/ATen/DLConvertor.cpp_. It also
introduces the following changes:

- Add a new Python API `torchDeviceToDLDevice()`, which is simply a
  refactoring of the `getDLDevice()` function at
  _aten/src/ATen/DLConvertor.cpp_.
- Add both keyword-arguments to the `from_dlpack()` function at
  _torch/utils/dlpack.py_ and to the `Tensor.__dlpack__()` dunder
  method.

[ghstack-poisoned]

ysiraichi committed:
Update on "[DLPack] Add support for missing keyword-arguments."


This PR introduces the rest of the keyword-arguments added in DLPack
version 2023.12: `dl_device` and `copy`.

In summary, we handle these arguments in the C++ implementation of
`to_dlpack(...)` at _torch/csrc/Module.cpp_, by calling the
`maybeCopyTensor` function at _aten/src/ATen/DLConvertor.cpp_. It also
introduces the following changes:

- Add a new Python API `torchDeviceToDLDevice()`, which is simply a
  refactoring of the `getDLDevice()` function at
  _aten/src/ATen/DLConvertor.cpp_.
- Add both keyword-arguments to the `from_dlpack()` function at
  _torch/utils/dlpack.py_ and to the `Tensor.__dlpack__()` dunder
  method.

[ghstack-poisoned]

ysiraichi committed:
Update on "[DLPack] Add support for missing keyword-arguments."


This PR introduces the rest of the keyword-arguments added in DLPack
version 2023.12: `dl_device` and `copy`.

In summary, we handle these arguments in the C++ implementation of
`to_dlpack(...)` at _torch/csrc/Module.cpp_, by calling the
`maybeCopyTensor` function at _aten/src/ATen/DLConvertor.cpp_. It also
introduces the following changes:

- Add a new Python API `torchDeviceToDLDevice()`, which is simply a
  refactoring of the `getDLDevice()` function at
  _aten/src/ATen/DLConvertor.cpp_.
- Add both keyword-arguments to the `from_dlpack()` function at
  _torch/utils/dlpack.py_ and to the `Tensor.__dlpack__()` dunder
  method.

[ghstack-poisoned]

ysiraichi committed:
Update on "[DLPack] Add support for missing keyword-arguments."


This PR introduces the rest of the keyword-arguments added in DLPack
version 2023.12: `dl_device` and `copy`.

In summary, we handle these arguments in the C++ implementation of
`to_dlpack(...)` at _torch/csrc/Module.cpp_, by calling the
`maybeCopyTensor` function at _aten/src/ATen/DLConvertor.cpp_. It also
introduces the following changes:

- Add a new Python API `torchDeviceToDLDevice()`, which is simply a
  refactoring of the `getDLDevice()` function at
  _aten/src/ATen/DLConvertor.cpp_.
- Add both keyword-arguments to the `from_dlpack()` function at
  _torch/utils/dlpack.py_ and to the `Tensor.__dlpack__()` dunder
  method.

[ghstack-poisoned]

albanD reviewed (approved):
I'm ok with skipping XPU handling until we hear back from them.
It might be a bit broken though, not sure if we have a way to test it?

guangyey commented on aten/src/ATen/DLConvertor.cpp:
```diff
@@ -388,4 +389,35 @@ Tensor fromDLPackVersioned(DLManagedTensorVersioned* src, std::function<void(voi
+  if (optional_dl_device.has_value()) {
+    auto device = at::getATenDevice(
+        optional_dl_device->device_type,
+        static_cast<c10::DeviceIndex>(optional_dl_device->device_id));
```
Yes, currently, XPU requires access to `data` within `getATenDevice`. @gujinghui could we refine this logic?

gujinghui replied:
Can we pass the data ptr here to not break xpu? We still need to check the device id upon the data ptr for a long term. Thanks.

ysiraichi commented:
@guangyey @gujinghui @albanD
After adding support for moving tensors onto XPU, I believe we need tests for it. So, I'm leaving it for a future PR.
I will probably try to merge this stack next Wednesday. So, if you think I should wait and fix this first, let me know until then.

ysiraichi committed:
Rebased. on "[DLPack] Add support for missing keyword-arguments."


This PR introduces the rest of the keyword-arguments added in DLPack
version 2023.12: `dl_device` and `copy`.

In summary, we handle these arguments in the C++ implementation of
`to_dlpack(...)` at _torch/csrc/Module.cpp_, by calling the
`maybeCopyTensor` function at _aten/src/ATen/DLConvertor.cpp_. It also
introduces the following changes:

- Add a new Python API `torchDeviceToDLDevice()`, which is simply a
  refactoring of the `getDLDevice()` function at
  _aten/src/ATen/DLConvertor.cpp_.
- Add both keyword-arguments to the `from_dlpack()` function at
  _torch/utils/dlpack.py_ and to the `Tensor.__dlpack__()` dunder
  method.

[ghstack-poisoned]

ysiraichi committed:
Update on "[DLPack] Add support for missing keyword-arguments."


This PR introduces the rest of the keyword-arguments added in DLPack
version 2023.12: `dl_device` and `copy`.

In summary, we handle these arguments in the C++ implementation of
`to_dlpack(...)` at _torch/csrc/Module.cpp_, by calling the
`maybeCopyTensor` function at _aten/src/ATen/DLConvertor.cpp_. It also
introduces the following changes:

- Add a new Python API `torchDeviceToDLDevice()`, which is simply a
  refactoring of the `getDLDevice()` function at
  _aten/src/ATen/DLConvertor.cpp_.
- Add both keyword-arguments to the `from_dlpack()` function at
  _torch/utils/dlpack.py_ and to the `Tensor.__dlpack__()` dunder
  method.

[ghstack-poisoned]

ysiraichi committed:
Update on "[DLPack] Add support for missing keyword-arguments."


This PR introduces the rest of the keyword-arguments added in DLPack
version 2023.12: `dl_device` and `copy`.

In summary, we handle these arguments in the C++ implementation of
`to_dlpack(...)` at _torch/csrc/Module.cpp_, by calling the
`maybeCopyTensor` function at _aten/src/ATen/DLConvertor.cpp_. It also
introduces the following changes:

- Add a new Python API `torchDeviceToDLDevice()`, which is simply a
  refactoring of the `getDLDevice()` function at
  _aten/src/ATen/DLConvertor.cpp_.
- Add both keyword-arguments to the `from_dlpack()` function at
  _torch/utils/dlpack.py_ and to the `Tensor.__dlpack__()` dunder
  method.

[ghstack-poisoned]

ysiraichi committed:
Update on "[DLPack] Add support for missing keyword-arguments."


This PR introduces the rest of the keyword-arguments added in DLPack
version 2023.12: `dl_device` and `copy`.

In summary, we handle these arguments in the C++ implementation of
`to_dlpack(...)` at _torch/csrc/Module.cpp_, by calling the
`maybeCopyTensor` function at _aten/src/ATen/DLConvertor.cpp_. It also
introduces the following changes:

- Add a new Python API `torchDeviceToDLDevice()`, which is simply a
  refactoring of the `getDLDevice()` function at
  _aten/src/ATen/DLConvertor.cpp_.
- Add both keyword-arguments to the `from_dlpack()` function at
  _torch/utils/dlpack.py_ and to the `Tensor.__dlpack__()` dunder
  method.

[ghstack-poisoned]
